src/task/future.rs

```diff
@@ -0,0 +1,393 @@
+}
+
+/// A [`Future`] task cell.
+pub struct TaskCell<Remote, Extras>(Arc<Task<Remote, Extras>>);
+
+// Safety: It is ensured that `future` and `extras` are always accessed by
+// only one thread at the same time.
+unsafe impl<Remote: Sync, Extras> Sync for Task<Remote, Extras> {}
+
+impl<Remote, Extras> fmt::Debug for TaskCell<Remote, Extras> {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        "future::TaskCell".fmt(f)
+    }
+}
+
+// When a future task is created or waken up by a waker, it is marked as
+// NOTIFIED. NOTIFIED tasks are ready to be polled. When the runner begins to
+// poll the future, it is marked as POLLING. When the runner finishes polling,
+// the future can either be ready or pending. If the future is ready, it is
+// marked as COMPLETED, or it checks whether it has becomes NOTIFIED. If it is
+// NOTIFIED, it should be polled again immediately. Otherwise it is marked as
+// IDLE.
+const NOTIFIED: u8 = 1;
+const IDLE: u8 = 2;
+const POLLING: u8 = 3;
+const COMPLETED: u8 = 4;
+
+impl<Remote, Extras> TaskCell<Remote, Extras> {
+    /// Creates a [`Future`] task cell that is ready to be polled.
+    pub fn new<F: Future<Output = ()> + Send + 'static>(
+        future: F,
+        remote: Remote,
+        extras: Extras,
+    ) -> Self {
+        TaskCell(Arc::new(Task {
+            status: AtomicU8::new(NOTIFIED),
+            future: UnsafeCell::new(Box::pin(future)),
+            remote,
+            extras: UnsafeCell::new(extras),
+        }))
+    }
+}
+
+impl<Remote, Extras> crate::queue::TaskCell for TaskCell<Remote, Extras> {
+    type Extras = Extras;
+
+    fn mut_extras(&mut self) -> &mut Self::Extras {
+        unsafe { &mut *self.0.extras.get() }
+    }
+}
+
+#[inline]
+unsafe fn waker<Remote, Extras>(task: *const Task<Remote, Extras>) -> Waker
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    Waker::from_raw(RawWaker::new(
+        task as *const (),
+        &RawWakerVTable::new(
+            clone_raw::<Remote, Extras>,
+            wake_raw::<Remote, Extras>,
+            wake_ref_raw::<Remote, Extras>,
+            drop_raw::<Remote, Extras>,
+        ),
+    ))
+}
+
+#[inline]
+unsafe fn clone_raw<Remote, Extras>(this: *const ()) -> RawWaker
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    let task_cell = clone_task(this as *const Task<Remote, Extras>);
+    RawWaker::new(
+        Arc::into_raw(task_cell.0) as *const (),
+        &RawWakerVTable::new(
+            clone_raw::<Remote, Extras>,
+            wake_raw::<Remote, Extras>,
+            wake_ref_raw::<Remote, Extras>,
+            drop_raw::<Remote, Extras>,
+        ),
+    )
+}
+
+#[inline]
+unsafe fn drop_raw<Remote, Extras>(this: *const ())
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    drop(task_cell(this as *const Task<Remote, Extras>))
+}
+
+unsafe fn wake_impl<Remote, Extras>(task_cell: &TaskCell<Remote, Extras>)
```

Comment:
How about taking `Cow<Arc<Task>>`? So that it can save one clone for `wake_up`.

Reply:
We can't. We need `remote` in the task to spawn. So even if we have the ownership of the task, we cannot spawn it with `remote` still in use. That means we need either cloning remote or task, which is both an atomic operation.